src/index.ts

```diff
@@ -1,178 +1,65 @@
-    this.configuration = {
-      ...this.configuration,
-      ...configuration,
-    };
-  };
-
-  get = async <ResponseType>({
-    path,
-    params,
-    options,
-  }: {
-    path: string;
-    params?: RequestParameters;
-    options?: RequestOptions;
-  }): Promise<ResponseType> => {
-    return ApiClient.returnData(
-      await fetch(this.getPath({ path, params, options }), {
-        method: 'GET',
-        headers: this.getHeaders(options),
-      })
-    );
-  };
-
-  post = async <ResponseType, RequestType>({
-    path,
-    params,
-    body,
-    options,
-  }: {
-    path: string;
-    params?: RequestParameters;
-    body: RequestType;
-    options?: RequestOptions;
-  }): Promise<ResponseType> => {
-    return ApiClient.returnData(
-      await fetch(this.getPath({ path, params, options }), {
-        method: 'POST',
-        headers: this.getHeaders(options),
-        body: body && JSON.stringify(body),
-      })
-    );
-  };
-
-  put = async <ResponseType, RequestType>({
-    path,
-    params,
-    body,
-    options,
-  }: {
-    path: string;
-    params?: RequestParameters;
-    body: RequestType;
-    options?: RequestOptions;
-  }): Promise<ResponseType> => {
-    return ApiClient.returnData(
-      await fetch(this.getPath({ path, params, options }), {
-        method: 'PUT',
-        headers: this.getHeaders(options),
-        body: body && JSON.stringify(body),
-      })
-    );
-  };
-
-  delete = async ({
-    path,
-    params,
-    options,
-  }: {
-    path: string;
-    params?: RequestParameters;
-    options?: RequestOptions;
-  }): Promise<void> => {
-    return ApiClient.returnData(
-      await fetch(this.getPath({ path, params, options }), {
-        method: 'DELETE',
-        headers: this.getHeaders(options),
-      })
-    );
-  };
+function StronglyTypedClient<Configuration extends ClientConfiguration>(): {
+  path: Path<Configuration>;
+} {
+  return {
+    path: (path, params) => {
+      const replacedPath = replaceParameters({
+        path: path.toString(),
+        params: params || {},
+      });
+      return {
+        get: ({ options } = { options: {} }) => {
+          return fetchClient.get({ path: replacedPath, options });
+        },
+        post: ({ data, options } = { data: {}, options: {} }) => {
+          return fetchClient.post({ path: replacedPath, body: data, options });
+        },
+        put: ({ data, options } = { data: {}, options: {} }) => {
+          return fetchClient.put({ path: replacedPath, body: data, options });
+        },
+        delete: ({ options } = { options: {} }) => {
+          return fetchClient.delete({ path: replacedPath, options });
+        },
+      };
+    },
+  } as { path: Path<Configuration> };
 }
 
-const apiClient = new ApiClient();
-
-export { ResponseError, apiClient as ApiClient };
+const initApiClient = (
+  configuration: Partial<FetchClientConfiguration> = {}
+) => {
+  fetchClient.init(configuration);
```

Comment:
See: https://github.com/carforyou/carforyou-listings-web/pull/4256/files#r879682284

Reply:
as discussed in slack, this would mean we are passing defaults like `baseUrl` and `headers` for each client. This would simplify mocking it in tests and gives more flexibility. I am ok with that. What do you think @Averethel ?

Reply:
So we would initialise every client with defaults instead of doing it globally? Do I understand correctly? Having good experience in mocking and tests would improve DX so I'm fine with this.